File: src/main/java/fr/epicanard/mapsaver/command/InfoCommand.java
package fr.epicanard.mapsaver.command;

import fr.epicanard.mapsaver.MapSaverPlugin;
import fr.epicanard.mapsaver.models.Permission;
import fr.epicanard.mapsaver.models.PlayerVisibility;
import fr.epicanard.mapsaver.models.language.MapInfo;
import fr.epicanard.mapsaver.models.map.PlayerAllMap;
import fr.epicanard.mapsaver.models.map.Visibility;
import fr.epicanard.mapsaver.utils.OptionalUtils;
import fr.epicanard.mapsaver.utils.PlayerUtils;
import fr.epicanard.mapsaver.utils.TextComponentBuilder;
import net.md_5.bungee.api.ChatColor;
import org.bukkit.Bukkit;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.MapMeta;

import java.util.List;
import java.util.Optional;

public class InfoCommand extends PlayerOnlyCommand {

    public InfoCommand(MapSaverPlugin plugin) {
        super(plugin, Permission.INFO_MAP, plugin.getLanguage().Help.Info);
    }

    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        final Optional<PlayerAllMap> playerAllMap = getPlayerAllMap(sender, args);

        playerAllMap
            .map(p -> {
                final MapInfo mapInfo = plugin.getLanguage().MapInfo;
                final Visibility visibility = p.getPlayerMap().getVisibility();
                final String ownerName = Bukkit.getOfflinePlayer(p.getPlayerMap().getPlayerUuid()).getName();
                return TextComponentBuilder.of()
                    .addLine("&7-------------------")
                    .addLine("&6%s : &f%s", mapInfo.Name, p.getPlayerMap().getName())
                    .addLine("&6%s : &f%s", mapInfo.Owner, ownerName)
                    .addLine("&6%s : &f%s", mapInfo.Visibility, plugin.getLanguage().Visibility.getOrDefault(visibility.name(), visibility.name()))
                    .addLine("&6%s : &f%d - %s", mapInfo.OriginalMap, p.getOriginalMap().getOriginalId().get(), p.getOriginalMap().getServer())
                    .addLine("&6%s :&f", mapInfo.CopyMaps)
                    .apply(builder -> p.getServerMaps().forEach(serverMap ->
                        builder.addLine(" • %d - %s", serverMap.getLockedId(), serverMap.getServer())
                    ))
                    .apply(builder -> {
                        if (Permission.IMPORT_MAP.isSetOn(sender)) {
                            builder
                                .prefix()
                                .add("&6%s : &f", mapInfo.Actions)
                                .addLink("import", plugin.getLanguage().List.ImportHover, ChatColor.DARK_GREEN, String.format("/mapsaver import %s %s", p.getPlayerMap().getName(), ownerName))
                            .bl();
                        }
                    })
                    .prefix().add("&7-------------------")
                    ;
            })
            .orElseGet(() -> TextComponentBuilder.of(plugin.getLanguage().ErrorMessages.MissingMapOrNotPublic))
            .send(sender);

        return true;
    }

    private Optional<PlayerAllMap> getPlayerAllMap(final CommandSender sender, final String[] args) {

        if (args.length == 0) {
            final PlayerVisibility playerVisibility = new PlayerVisibility(
                ((Player) sender).getUniqueId(),
                OptionalUtils.when(() -> !Permission.ADMIN_INFO_MAP.isSetOn(sender), Visibility.PUBLIC)
            );
            final ItemStack mapItem = ((Player) sender).getInventory().getItemInMainHand();
            return Optional.ofNullable((MapMeta) mapItem.getItemMeta())
                .map(MapMeta::getMapView)
                .flatMap(mapView -> plugin.getService().getMapInfo(playerVisibility, mapView.getId()));
        }

        final String playerName = (args.length >= 2) ? args[1] : null;
        final PlayerVisibility playerVisibility = PlayerUtils.extractPlayerVisibility(plugin, playerName, sender, Permission.ADMIN_INFO_MAP);
        return plugin.getService().getMapInfo(playerVisibility, args[0]);
    }

    public boolean isPlayerOnly(final CommandSender sender, final String[] args) {
        return (args.length >= 2) || super.isPlayerOnly(sender, args);
    }


    @Override
    public List<String> onTabComplete(CommandSender commandSender, Command command, String s, String[] strings) {
        return null;
    }
}

File: src/main/scala/fr/epicanard/mapsaver/commands/InfoCommand.scala
package fr.epicanard.mapsaver.commands

import cats.data.EitherT
import fr.epicanard.mapsaver.Permission
import fr.epicanard.mapsaver.commands.CommandContext.getPlayer
import fr.epicanard.mapsaver.commands.InfoCommand.{buildMessage, getPlayerServerMaps}
import fr.epicanard.mapsaver.database.MapRepository
import fr.epicanard.mapsaver.errors.Error
import fr.epicanard.mapsaver.map.MapExtractor
import fr.epicanard.mapsaver.message.Message._
import fr.epicanard.mapsaver.message.{Component, Message, Messenger}
import fr.epicanard.mapsaver.models.map.{PlayerServerMaps, Visibility}
import fr.epicanard.mapsaver.models.{Complete, Player, RestrictVisibility}
import fr.epicanard.mapsaver.resources.language.{Help, Language, Visibilities}
import net.md_5.bungee.api.ChatColor
import org.bukkit.OfflinePlayer
import org.bukkit.command.CommandSender

import scala.concurrent.{ExecutionContext, Future}

case class InfoCommand(mapRepository: MapRepository)(implicit ec: ExecutionContext)
    extends BaseCommand(Some(Permission.ListMap)) {

  def helpMessage(help: Help): String = help.info

  def onCommand(messenger: Messenger, commandContext: CommandContext): Future[Either[Error, Message]] =
    (for {
      playerServerMaps <- EitherT(getPlayerServerMaps(mapRepository, commandContext))
      message = buildMessage(messenger.language, playerServerMaps, commandContext.sender)
    } yield message).value

  def onTabComplete(commandContext: CommandContext): Future[Either[Error, Complete]] =
    BaseCommand.mapTabComplete(mapRepository, commandContext)
}

object InfoCommand {
  private case class MapsWithOwner(playerServerMaps: PlayerServerMaps, owner: OfflinePlayer)

  private def getPlayerServerMaps(
      mapRepository: MapRepository,
      commandContext: CommandContext
  )(implicit ec: ExecutionContext): Future[Either[Error, MapsWithOwner]] =
    commandContext.args match {
      case Nil                 => getInfoMapInHand(mapRepository, commandContext)
      case name :: Nil         => getInfoOfSender(mapRepository, commandContext, name)
      case player :: name :: _ => getInfoOfPlayer(mapRepository, commandContext, name, player)
    }

  private def getInfoMapInHand(
      mapRepository: MapRepository,
      commandContext: CommandContext
  )(implicit ec: ExecutionContext): Future[Either[Error, MapsWithOwner]] =
    (for {
      player <- EitherT.fromEither[Future](getPlayer(commandContext))
      restrictVisibility = RestrictVisibility.unless(Permission.AdminInfoMap, player)(RestrictVisibility.OwnerOrPublic)
      mapView <- EitherT.fromEither[Future](MapExtractor.extractMapView(player))
      playerServerMaps <- EitherT(
        mapRepository
          .getMapInfo(player.getUniqueId, restrictVisibility, mapView.getId, commandContext.server)
      )
      owner = Player.getOfflinePlayer(playerServerMaps.playerMap.playerUuid)
    } yield MapsWithOwner(playerServerMaps, owner)).value

  private def getInfoOfSender(
      mapRepository: MapRepository,
      commandContext: CommandContext,
      name: String
  )(implicit ec: ExecutionContext): Future[Either[Error, MapsWithOwner]] =
    (for {
      player           <- EitherT.fromEither[Future](getPlayer(commandContext))
      playerServerMaps <- EitherT(mapRepository.getMapInfo(player.getUniqueId, None, name))
    } yield MapsWithOwner(playerServerMaps, player)).value

  private def getInfoOfPlayer(
      mapRepository: MapRepository,
      commandContext: CommandContext,
      name: String,
      playerName: String
  )(implicit ec: ExecutionContext): Future[Either[Error, MapsWithOwner]] = {
    val owner              = Player.getOfflinePlayer(playerName)
    val restrictVisibility = Visibility.getRestrictVisibility(commandContext, owner, Permission.AdminInfoMap)
    EitherT(mapRepository.getMapInfo(owner.getUniqueId, restrictVisibility, name))
      .map(maps => MapsWithOwner(maps, owner))
      .value
  }

  private def buildMessage(
      language: Language,
      mapsWithOwner: MapsWithOwner,
      sender: CommandSender
  ): Message = {
    val mapInfo          = language.mapInfo
    val playerServerMaps = mapsWithOwner.playerServerMaps
    val visibility       = Visibilities.fromVisibility(playerServerMaps.playerMap.visibility, language.visibilities)
    msg"""&7-------------------
       |&6${mapInfo.name} : &f${playerServerMaps.playerMap.name}
       |&6${mapInfo.owner} : &f${mapsWithOwner.owner.getName}
       |&6${mapInfo.visibility} : &f$visibility
       |&6${mapInfo.originalMap} : &f${playerServerMaps.originalMap.originalId.get} - ${playerServerMaps.originalMap.server}#${playerServerMaps.originalMap.world} - [${playerServerMaps.originalMap.x},${playerServerMaps.originalMap.z}]
       |&6${mapInfo.copyMaps} : &f""" ++
      Message(playerServerMaps.serverMaps.foldLeft(List[Component]()) { case (comps, serverMap) =>
        comps :+ Component(s" • ${serverMap.lockedId} - ${serverMap.server}")
      }) +?
      Option.when(Permission.ImportMap.isSetOn(sender, defaultSender = false))(
        Component(s"&6${mapInfo.actions} : &f") + Component.link(
          "import",
          language.list.importHover,
          ChatColor.DARK_GREEN,
          s"""/mapsaver import ${mapsWithOwner.owner.getName} "${playerServerMaps.playerMap.name}""""
        )
      ) +?
      Option.when(playerServerMaps.playerMap.locked)(Component(s"&c${mapInfo.locked}")) ++
      msg"""&7-------------------"""
  }
}
